### ISSUE TITLE
ENH: ndimage.generic_filter: support complex input

### ISSUE THREAD

**Issue by yeewantung** [scipy/scipy#18437]
https://github.com/scipy/scipy/blob/91210d7adc77cbeaa23023c7ab1185fda8e95009/scipy/ndimage/_filters.py#L1658

I define a function which operators on arrays of complex type and wish to use generic_filter to batch apply that function to my input arrays. It is disappointing that generic_filter itself doesn't support complex type. The generic_filter should be able to operate on complex type as long as the function itself can tolerate complex type.

**Comment by grlee77** [comment]
I agree that this would ideally be supported, but it is not trivial to due to the [function signature](https://github.com/scipy/scipy/blob/0c590f4d74ab7d6f41426c7717204c0edb662d73/scipy/ndimage/src/ni_filters.c#L943-L946) of the wrapped C function expecting a function that takes `double*` data and across the C code buffers tend to get allocated using `double` for any temporary buffers. It would unfortunately require a substantial refactoring of the C code to enable.
